src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs

```diff
@@ -1506,28 +1510,92 @@ HashSet<string> GetPackageDependencies()
+                        var packageReverseDependencies = new Dictionary<string, HashSet<string>>(_compileTimeTarget.Libraries.Count, StringComparer.OrdinalIgnoreCase);
+                        for (int i = 0; i < _compileTimeTarget.Libraries.Count; i++)
+                        {
+                            var parentPackage = _compileTimeTarget.Libraries[i];
+                            if (string.IsNullOrEmpty(parentPackage.Name)) continue;
+
+                            if (!packageReverseDependencies.ContainsKey(parentPackage.Name))
+                            {
+                                packageReverseDependencies[parentPackage.Name] = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                            }
+
+                            for (int j = 0; j < parentPackage.Dependencies.Count; j++)
+                            {
+                                var dependency = parentPackage.Dependencies[j].Id;
+
+                                if (!packageReverseDependencies.TryGetValue(dependency, out HashSet<string> parentPackages))
+                                {
+                                    parentPackages = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                                    packageReverseDependencies[dependency] = parentPackages;
+                                }
+
+                                parentPackages.Add(parentPackage.Name);
+                            }
+                        }
+                        return packageReverseDependencies;
+                    }
+
+                    bool ForCurrentTargetFramework(string targetFramework)
+                    {
+                        var parts = targetFramework.Split(LockFile.DirectorySeparatorChar);
+                        var parsedTargetGraph = NuGetFramework.Parse(parts[0]);
+                        var alias = _lockFile.PackageSpec.TargetFrameworks
+                            .FirstOrDefault(tf => tf.FrameworkName == parsedTargetGraph)
+                            ?.TargetAlias ?? targetFramework;
+                        return alias == _task.TargetFramework;
                     }
 
-                    return messages.Max(log => log.Level).ToString();
+                    void ApplyDiagnosticLevel(string package, LogLevel logLevel, Dictionary<string, LogLevel> diagnosticLevels, Dictionary<string, HashSet<string>> reverseDependencies)
+                    {
+                        if (!reverseDependencies.TryGetValue(package, out HashSet<string> parentPackages))
+                        {
+                            // The package is not used in the current TargetFramework
+                            return;
+                        }
+
+                        if (diagnosticLevels.TryGetValue(package, out LogLevel currentLevel))
+                        {
+                            // Only continue if we need to increase the level
+                            if (currentLevel <= logLevel)
```

Comment:
Is this check correct? What about the following scenario?

* Package A, Lowest Diagnostic Level
 - Dependency 1, Lowest Diagnostic Level
 -> - Dependency 2 - HAS VULNERABILITY
 -> -> - Dependency 3, Lowest Diagnostic Level

If we start at 1 or 3, which have the same diagnostic level as the parent/root, will this not return out and miss the diagnostic level of dependency 2? I'm not very familiar with this space, but I'd assume dependency 2's diagnostic level would not

Reply:
good catch!  I should have written a better test the first time.  I updated the test to cover this scenario now.